### ISSUE TITLE
TerminatingWithPathSpecimenBuilder design decision

### ISSUE THREAD

**Issue by zvirja** [AutoFixture/AutoFixture#771]
I'm looking into way to improve the #694 and want to clarify the design decision for the `TerminatingWithPathSpecimenBuilder`. The question is why we use the inner `TracingBuilder` and its events instead of working with stack directly in the `Create()` method? 

Currently, we register `TerminatingWithPathSpecimenBuilder` as following:
```csharp
new TerminatingWithPathSpecimenBuilder(new TracingBuilder(new CompositeSpecimenBuilder(
...
```

Later we subscribe to TracingBuilder events [to manage our stack](https://github.com/AutoFixture/AutoFixture/blob/a6de0ce7b93707a34d8db063d4506126bc8a1dea/Src/AutoFixture/Kernel/TerminatingWithPathSpecimenBuilder.cs#L44-L54). If specimen builder wasn't created, we throw exception.

The design is quite complicated as we closely interact with the inner `TracingBuilder` to manage our local stack. Also, if exception is thrown, the most inner `TerminatingWithPathSpecimenBuilder` is responsible for the stack cleanup - which isn't too elegant design.

Instead, it might be refactored so something like:
```csharp
public object Create(object request, ISpecimenContext context)
{
    GetPathForCurrentThread().Push(request);
    try
    {
        var result = this.Builder.Create(request, context);
        if (result is NoSpecimen)
        {
            throw new ObjectCreationException(string.Format(
                CultureInfo.CurrentCulture,
                BuildCoreMessageTemplate(request),
                request,
                Environment.NewLine,
                BuildRequestPathText(this.SpecimenRequests)));
        }
    }
    finally
    {
        GetPathForCurrentThread().Pop();
    }
}
```

It would be much easier and will read much simpler. 

I do understand that here might be a couple of reasons: 
- We found that we could re-use some part of the `TracingBuilder` logic.
- We wanted to add possibility to perform filtration of which requests are logged and re-used the `TracingBuilder` abstraction for that.

Given the nature of this handler (it produces results in exceptional situations only), I'd say that we could safely simplify it - remove inner `TracingBuilder`.

Most likely, that is a question to the @adamchester as he created it.

**Comment by adamchester** [comment]
I agree with you @zvirja, the current implementation is quite complicated to reason about, especially the way it interacts with the inner `TracingBuilder`.

It looks like what you have proposed will improve things a lot. 👍

**Comment by zvirja** [comment]
@adamchester Thanks for the reply! If I remove the tracing builder and use the approach I described, that will:
- require some tests to be altered/deleted;
- will remove the Tracing Builder from the Fixture graph. 

Do I see that right and such changes are acceptable for the v4 branch only?

**Comment by adamchester** [comment]
I think if you modify the public behaviour of in the way you described above, then yeah it needs to be in v4 only.

I'm not 100% sure if there's a way to introduce a new path-building-terminating mechanism without breaking compatibility but it seems worth exploring?

**Comment by zvirja** [comment]
I'd like to incorporate changes related to the #694, so it should worth it.

@moodmosaic and @ecampidoglio Do you have any concerns regarding the idea of simplifying the logic?

**Comment by moodmosaic** [comment]
Sorry, I don't know how these parts fit together, so I hope @adamchester or @ecampidoglio can be of help.

**Comment by zvirja** [comment]
Closed by #787
